### PR TITLE
Moving tests from test/ to python/ray/tests/

### PR DESCRIPTION
**Pull request by williamma12**
## What do these changes do?

Moves relevant python tests to python/ray/test/ so that it will ship with ray when installed from Pypi.

### CODE REVIEW

**Comment by williamma12** [comment]
Using this comment to start a discussion on the motivations behind why certain tests were moved and others were not

Moved:
* `actor_test.py`: mainly just tests that the python api passes the arguments and other common class functionality to ray properly
* `array_test.py`: Tests `ray.experimental.array` functions work as expected
* `cython_test.py`: Tests that cython works as expected with ray
* `failure_test.py`: Tests that ray propagates errors correctly
* `mini_test.py`: Tests basic ray python api calls
* `recursion_test.py`: Tests recursive python functions work as expected
* `runtest.py`: Tests various ray’s python api functionality
* `tensorflow_test.py` Tests that tensorflow works as expected


Not Moved:
* `autoscaler.py`: Small stress test
* `component_failures_test.py`:Fault tolerance testing for each of the different parts of ray
* `credits_test.py`: Tests redis
* `debug_tools_test.py`: Tests that the debug tools still work
* `microbenchmarks.py`: Benchmarking
* `jenkins_tests/`: Stress tests
* `monitor_test.py`: Tests that the monitor returns the expected values
* `multi_node_test.py`: Tests that the nodes behave as expected
* `multi_node_test_2.py`: More tests that tests that nodes behave as expected
* `node_manager_test.py`: Tests the node manager works for edge cases
* `object_manager_test.py`: Tests that the object manager works as expected
* `stress_tests/` and `stress_tests.py`: Stress tests
* `tempfile_test.py`: Tests that the temporary files are created properly

**Comment by robertnishihara** [comment]
Looks like it should be `tests` not `test`. See https://python-packaging.readthedocs.io/en/latest/testing.html.

**Comment by robertnishihara** [comment]
Thanks @williamma12! I think we should actually move all of the tests except `stress_tests/*` and `jenkins_tests/`.

**Comment by robertnishihara** [comment]
@pcmoritz what do you think?

**Comment by pcmoritz** [comment]
I agree and we can also rename the "test" directory in the root to "ci" or "integration" something else so it is clear that tests should not go there, but into the new folder instead.

**Comment by williamma12** [comment]
Sounds good! I've moved all the files from the old `test/` to `python/ray/test` except for `stress_tests/`, `jenkins_tests/`, and `microbenchmarking.py`. `microbenchmarking.py` was not moved because it does not actually contain any pytest tests but just benchmarks that prints the benchmarks. 

I have also updated the folder names from `test/` and `python/ray/test/` to `integration/` and `python/ray/tests`, respectively. 

Also, I am open to suggestions for better names for the new `integration/` folder!

File: ci/jenkins_tests/multi_node_tests/many_drivers_test.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 import time
 
 import ray
-from ray.test.test_utils import (_wait_for_nodes_to_join, _broadcast_event,
+from ray.tests.test_utils import (_wait_for_nodes_to_join, _broadcast_event,
```
Comment (robertnishihara) [inline ci/jenkins_tests/multi_node_tests/many_drivers_test.py]: rename to `utils`

File: .travis.yml [hunk 2]
```diff
@@ -172,41 +172,41 @@ script:
...
-  - python -m pytest -v --durations=10 test/tempfile_test.py
+  - python -m pytest -v --durations=10 python/ray/tests/test_tempfile.py
 
   # ray debug tools tests
-  - python -m pytest -v --durations=10 test/debug_tools_test.py
+  - python -m pytest -v --durations=10 python/ray/tests/test_debug_tools.py
 
   # modin test files
-  - python python/ray/test/test_modin.py
+  - python python/ray/tests/test_modin.py
```
Comment (raulchen) [inline .travis.yml]: I think now it's a good chance to follow the test discovery convention (see https://docs.pytest.org/en/latest/goodpractices.html#conventions-for-python-test-discovery) and change these file-based commands to to one single directory-based command (or a few sub-directories if needed).
The benefit is that we can only have one statement in this travis ci file, e.g., `pytest -v --duration=10 python/ray/tests`. And when someone adds a new test file, they don't need to modify here.

Comment (williamma12) [reply .travis.yml]: Thanks! Doing so also revealed that the `test_queue.py` lacked a `teardown_module`, which I added in #4012. 

I notice that there was also a comment about `object_manager_test.py` and I left it as a separate command but I am open to suggestions

File: .travis.yml [hunk 3]
```diff
@@ -172,41 +172,13 @@ script:
...
-
-  # ray temp file tests
-  - python -m pytest -v --durations=10 test/tempfile_test.py
-
-  # ray debug tools tests
-  - python -m pytest -v --durations=10 test/debug_tools_test.py
-
-  # modin test files
-  - python python/ray/test/test_modin.py
+  - python -m pytest -v --durations=10 python/ray/tests/test_object_manager.py
```
Comment (raulchen) [inline .travis.yml]: I think this will run `test_object_manager.py` twice, right? Maybe we can move the above comment to the test_object_manager.py file, and don't need to make this a separate command. cc @guoyuhong

File: ci/travis/install-dependencies.sh [hunk 4]
```diff
@@ -24,7 +24,7 @@ if [[ "$PYTHON" == "2.7" ]] && [[ "$platform" == "linux" ]]; then
   wget https://repo.continuum.io/miniconda/Miniconda2-4.5.4-Linux-x86_64.sh -O miniconda.sh -nv
   bash miniconda.sh -b -p $HOME/miniconda
   export PATH="$HOME/miniconda/bin:$PATH"
-  pip install -q cython==0.29.0 cmake tensorflow gym==0.10.11 opencv-python pyyaml pandas==0.23.4 requests \
```
Comment (robertnishihara) [inline ci/travis/install-dependencies.sh]: Is this change intentional? I think we want to pin `gym` for Python 2.7 until the next release. See the discussion in https://github.com/ray-project/ray/pull/3973.

Comment (williamma12) [reply ci/travis/install-dependencies.sh]: oops I did not notice that change. I should have undid it

File: ci/travis/install-dependencies.sh [hunk 5]
```diff
@@ -50,7 +50,7 @@ elif [[ "$PYTHON" == "2.7" ]] && [[ "$platform" == "macosx" ]]; then
   wget https://repo.continuum.io/miniconda/Miniconda2-4.5.4-MacOSX-x86_64.sh -O miniconda.sh -nv
   bash miniconda.sh -b -p $HOME/miniconda
   export PATH="$HOME/miniconda/bin:$PATH"
-  pip install -q cython==0.29.0 cmake tensorflow gym==0.10.11 opencv-python pyyaml pandas==0.23.4 requests \
```
Comment (robertnishihara) [inline ci/travis/install-dependencies.sh]: Undo this.

File: .travis.yml (hunk 3 above)
Comment (williamma12) [reply .travis.yml]: I reworded the comment so that I could keep the comment in `.travis.yml` and delete the line so `test_object_manager.py` won't be ran twice

**Comment by robertnishihara** [comment]
@williamma12 this is failing linting and it looks like a bunch of other tests as well.

**Comment by robertnishihara** [comment]
Check if the change to `test_recursion.py` works locally because I've seen issues when running that in pytest in the past.

File: python/ray/tests/test_recursion.py [hunk 6]
```diff
@@ -6,19 +6,30 @@
...
+    # Shutdown ray instance
+    ray.shutdown()
 
-@ray.remote
-def factorial(n):
-    if n == 0:
-        return 1
-    return n * ray.get(factorial.remote(n - 1))
 
+def test_recursion(init):
```
Comment (robertnishihara) [inline python/ray/tests/test_recursion.py]: @williamma12 does this test actually work in pytest?

Comment (williamma12) [reply python/ray/tests/test_recursion.py]: oops good catch. Did not intend to move the remote function inside the test and undoing it made the test pass.

Comment (robertnishihara) [reply python/ray/tests/test_recursion.py]: For consistency, can you rename `init` to `ray_start`?

**Comment by robertnishihara** [comment]
There seems to be an issue with `test_monitor.py`.

**Review by robertnishihara — APPROVED** [review]
LGTM once tests pass.

**Comment by robertnishihara** [comment]
Jenkins, retest this please.
